anomdetect/src/main/java/com/expedia/adaptivealerting/anomdetect/DetectorMapper.java

```diff
@@ -15,43 +15,90 @@
  */
 package com.expedia.adaptivealerting.anomdetect;
 
+
 import com.expedia.adaptivealerting.anomdetect.comp.DetectorSource;
+import com.expedia.adaptivealerting.anomdetect.mapper.CacheUtil;
+import com.expedia.adaptivealerting.anomdetect.mapper.Detector;
+import com.expedia.adaptivealerting.anomdetect.mapper.DetectorMatchResponse;
 import com.expedia.adaptivealerting.core.data.MappedMetricData;
 import com.expedia.metrics.MetricData;
 import lombok.Getter;
 import lombok.NonNull;
-import lombok.RequiredArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
 
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
 import java.util.Set;
-import java.util.stream.Collectors;
-
-import static com.expedia.adaptivealerting.core.util.AssertUtil.notNull;
+import java.util.concurrent.atomic.AtomicLong;
 
 /**
  * Entry into the Adaptive Alerting runtime. Its job is find for any incoming {@link MetricData} the corresponding set
  * of mapped detectors, creating a {@link MappedMetricData} for each.
  */
-@RequiredArgsConstructor
 @Slf4j
 public class DetectorMapper {
-
+    private DetectorMapperCache cache;
+    private AtomicLong lastElasticLookUpLatency = new AtomicLong(-1);
     @Getter
     @NonNull
     private DetectorSource detectorSource;
 
-    /**
-     * Maps an {@link MetricData} to its corresponding set of {@link MappedMetricData}s.
-     *
-     * @param metricData MetricData to map.
-     * @return The corresponding set of {@link MappedMetricData}s: one per detector.
-     */
-    public Set<MappedMetricData> map(MetricData metricData) {
-        notNull(metricData, "metricData can't be null");
-        return detectorSource
-                .findDetectorUuids(metricData.getMetricDefinition())
-                .stream()
-                .map(detectorUuid -> new MappedMetricData(metricData, detectorUuid))
-                .collect(Collectors.toSet());
+
+    public DetectorMapper(DetectorSource detectorSource) {
+        assert detectorSource != null;
+        this.detectorSource = detectorSource;
+        this.cache = new DetectorMapperCache();
+    }
+
+    public List<Detector> getDetectorsFromCache(MetricData metricData) {
```

Comment:
I wonder whether

```
public List<String> getDetectorUuids(MetricDefinition metricDef)
```

wouldn't be a better signature here. Several reasons:

- I don't see a reason to expose the fact that `DetectorMapper` uses a cache at all. Seems like an implementation detail. Couldn't we make cache management (including evicting stale entries) internal to the `DetectorMapper`?
- With respect to `Detector`, it contains only a UUID, and the only thing the lookup cares about is finding UUIDs. (There is no "detector type" concept anymore so we don't have to worry about that.) So is there a point to the `Detector` abstraction at all?
- Requiring a `MetricData` here feels a little iffy here. Conceptually it's the `MetricDefinition` that serves as the metric identity, and using that as the parameter makes it clearer that we're looking up the detector UUIDs for a given metric.

Reply:
(Not totally sure about the idea that we don't expose the cache. I do understand that this component needs to work with something that notices detector changes. Treat this as a discussion more than a request.)

Reply:
Keeping in Detector abstraction, might make it easier to modify this incase we want to utilise other detector related info in future.